## Ejercicio: Análisis Exploratorio de Datos

### 1. Objetivos:
    - Practicar leer archivos JSON usando pandas
    - Practicar hacerse preguntas acerca de los conjuntos de datos que tenemos

---
    
### 2. Desarrollo:

Vamos a practicar explorar conjuntos de datos y hacernos preguntas acerca de ellos.

Tenemos un conjunto de datos en formato JSON 'new_york_times_bestsellers-clean.json' (en Classroom).

Primero que nada, lee el archivo JSON y crea un `DataFrame` con él:

In [12]:
## Realiza aquí los imports que necesites
import pandas as pd
## Lee aquí tu archivo JSON
df = pd.read_json('new_york_times_bestsellers-clean.json')
## Crea aquí tu DataFrame
df

amazon_product_url  \
0     http://www.amazon.com/The-Host-Novel-Stephenie...   
1     http://www.amazon.com/Love-Youre-With-Emily-Gi...   
2     http://www.amazon.com/The-Front-Garano-Patrici...   
3     http://www.amazon.com/Snuff-Chuck-Palahniuk/dp...   
4     http://www.amazon.com/Sundays-at-Tiffanys-Jame...   
...                                                 ...   
3028  http://www.amazon.com/Six-Years-Harlan-Coben/d...   
3029  http://www.amazon.com/The-Interestings-Novel-M...   
3030  http://www.amazon.com/Man-Without-Breath-Berni...   
3031  http://www.amazon.com/The-Storyteller-Jodi-Pic...   
3032  http://www.amazon.com/Z-A-Novel-Zelda-Fitzgera...   

                                        author  \
0                              Stephenie Meyer   
1                                 Emily Giffin   
2                            Patricia Cornwell   
3                              Chuck Palahniuk   
4     James Patterson and Gabrielle Charbonnet   
...                                        ...   
3028                              Harlan Coben   
3029                              Meg Wolitzer   
3030                               Philip Kerr   
3031                              Jodi Picoult   
3032                       Therese Anne Fowler   

                                            description            publisher  \
0     Aliens have taken control of the minds and bod...        Little, Brown   
1     A woman's happy marriage is shaken when she en...         St. Martin's   
2     A Massachusetts state investigator and his tea...               Putnam   
3     An aging porn queens aims to cap her career by...            Doubleday   
4                      A woman finds an unexpected love        Little, Brown   
...                                                 ...                  ...   
3028  Jake Fisher discovers that neither the woman h...               Dutton   
3029  Six friends meet in the 1970s at a summer arts...            Riverhead   
3030  Bernie Gunther, the Berlin cop, is sent to Smo...   Marian Wood/Putnam   
3031  A New Hampshire baker finds herself in the mid...  Emily Bestler/Atria   
3032  A novel based on the lives of Zelda and F. Sco...         St. Martin's   

                         title                       oid  \
0                     THE HOST  5b4aa4ead3089013507db18c   
1     LOVE THE ONE YOU'RE WITH  5b4aa4ead3089013507db18d   
2                    THE FRONT  5b4aa4ead3089013507db18e   
3                        SNUFF  5b4aa4ead3089013507db18f   
4         SUNDAYS AT TIFFANY’S  5b4aa4ead3089013507db190   
...                        ...                       ...   
3028                 SIX YEARS  5b4aa4ead3089013507dc593   
3029          THE INTERESTINGS  5b4aa4ead3089013507dc595   
3030      A MAN WITHOUT BREATH  5b4aa4ead3089013507dc597   
3031           THE STORYTELLER  5b4aa4ead3089013507dc598   
3032                         Z  5b4aa4ead3089013507dc59a   

      bestsellers_date.numberLong  published_date.numberLong  rank.numberInt  \
0                   1211587200000              1212883200000               2   
1                   1211587200000              1212883200000               3   
2                   1211587200000              1212883200000               4   
3                   1211587200000              1212883200000               5   
4                   1211587200000              1212883200000               6   
...                           ...                        ...             ...   
3028                1366416000000              1367712000000               9   
3029                1366416000000              1367712000000              11   
3030                1366416000000              1367712000000              13   
3031                1366416000000              1367712000000              14   
3032                1366416000000              1367712000000              16   

      rank_last_week.numberInt  weeks_on_list.numberInt  price.numberDouble  
0     

Ahora, usando todas las herramientas que hemos aprendido en esta sesión (indexación de filas y columnas, `shape`, `dtypes`, `head`, `tail`, `columns`, `info`, etc) explora tu dataset.

In [4]:
df.shape

(3033, 12)

In [5]:
df.head(5)

amazon_product_url  \
0  http://www.amazon.com/The-Host-Novel-Stephenie...   
1  http://www.amazon.com/Love-Youre-With-Emily-Gi...   
2  http://www.amazon.com/The-Front-Garano-Patrici...   
3  http://www.amazon.com/Snuff-Chuck-Palahniuk/dp...   
4  http://www.amazon.com/Sundays-at-Tiffanys-Jame...   

                                     author  \
0                           Stephenie Meyer   
1                              Emily Giffin   
2                         Patricia Cornwell   
3                           Chuck Palahniuk   
4  James Patterson and Gabrielle Charbonnet   

                                         description      publisher  \
0  Aliens have taken control of the minds and bod...  Little, Brown   
1  A woman's happy marriage is shaken when she en...   St. Martin's   
2  A Massachusetts state investigator and his tea...         Putnam   
3  An aging porn queens aims to cap her career by...      Doubleday   
4                   A woman finds an unexpected love  Little, Brown   

                      title                       oid  \
0                  THE HOST  5b4aa4ead3089013507db18c   
1  LOVE THE ONE YOU'RE WITH  5b4aa4ead3089013507db18d   
2                 THE FRONT  5b4aa4ead3089013507db18e   
3                     SNUFF  5b4aa4ead3089013507db18f   
4      SUNDAYS AT TIFFANY’S  5b4aa4ead3089013507db190   

   bestsellers_date.numberLong  published_date.numberLong  rank.numberInt  \
0                1211587200000              1212883200000               2   
1                1211587200000              1212883200000               3   
2                1211587200000              1212883200000               4   
3                1211587200000              1212883200000               5   
4                1211587200000              1212883200000               6   

   rank_last_week.numberInt  weeks_on_list.numberInt  price.numberDouble  
0                         1                        3               25.99  
1                         2                        2               24.95  
2                         0                        1               22.95  
3                         0                        1               24.95  
4                         3                        4               24.99

In [6]:
df.tail(5)

amazon_product_url               author  \
3028  http://www.amazon.com/Six-Years-Harlan-Coben/d...         Harlan Coben   
3029  http://www.amazon.com/The-Interestings-Novel-M...         Meg Wolitzer   
3030  http://www.amazon.com/Man-Without-Breath-Berni...          Philip Kerr   
3031  http://www.amazon.com/The-Storyteller-Jodi-Pic...         Jodi Picoult   
3032  http://www.amazon.com/Z-A-Novel-Zelda-Fitzgera...  Therese Anne Fowler   

                                            description            publisher  \
3028  Jake Fisher discovers that neither the woman h...               Dutton   
3029  Six friends meet in the 1970s at a summer arts...            Riverhead   
3030  Bernie Gunther, the Berlin cop, is sent to Smo...   Marian Wood/Putnam   
3031  A New Hampshire baker finds herself in the mid...  Emily Bestler/Atria   
3032  A novel based on the lives of Zelda and F. Sco...         St. Martin's   

                     title                       oid  \
3028             SIX YEARS  5b4aa4ead3089013507dc593   
3029      THE INTERESTINGS  5b4aa4ead3089013507dc595   
3030  A MAN WITHOUT BREATH  5b4aa4ead3089013507dc597   
3031       THE STORYTELLER  5b4aa4ead3089013507dc598   
3032                     Z  5b4aa4ead3089013507dc59a   

      bestsellers_date.numberLong  published_date.numberLong  rank.numberInt  \
3028                1366416000000              1367712000000               9   
3029                1366416000000              1367712000000              11   
3030                1366416000000              1367712000000              13   
3031                1366416000000              1367712000000              14   
3032                1366416000000              1367712000000              16   

      rank_last_week.numberInt  weeks_on_list.numberInt  price.numberDouble  
3028                         8                        5               27.95  
3029                        11                        2               27.95  
3030                         0                        1               26.95  
3031                        10                        8               28.99  
3032                         0                        3               25.99

In [7]:
df.dtypes

amazon_product_url              object
author                          object
description                     object
publisher                       object
title                           object
oid                             object
bestsellers_date.numberLong      int64
published_date.numberLong        int64
rank.numberInt                   int64
rank_last_week.numberInt         int64
weeks_on_list.numberInt          int64
price.numberDouble             float64
dtype: object

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3033 entries, 0 to 3032
Data columns (total 12 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   amazon_product_url           3033 non-null   object 
 1   author                       3033 non-null   object 
 2   description                  3033 non-null   object 
 3   publisher                    3033 non-null   object 
 4   title                        3033 non-null   object 
 5   oid                          3033 non-null   object 
 6   bestsellers_date.numberLong  3033 non-null   int64  
 7   published_date.numberLong    3033 non-null   int64  
 8   rank.numberInt               3033 non-null   int64  
 9   rank_last_week.numberInt     3033 non-null   int64  
 10  weeks_on_list.numberInt      3033 non-null   int64  
 11  price.numberDouble           3033 non-null   float64
dtypes: float64(1), int64(5), object(6)
memory usage: 308.0+ KB


In [9]:
df.columns

Index(['amazon_product_url', 'author', 'description', 'publisher', 'title',
       'oid', 'bestsellers_date.numberLong', 'published_date.numberLong',
       'rank.numberInt', 'rank_last_week.numberInt', 'weeks_on_list.numberInt',
       'price.numberDouble'],
      dtype='object')

debate con el experto y tus compañeros las siguientes preguntas:

1. ¿Qué podemos saber de este dataset con tan sólo leer el nombre del archivo y los nombres de las columnas?

R: Es una lista de los libros más vendidos con sus respectivos atributos como autor, editorial así como los indicadores para decir que está en el top o estuvo.


2. ¿Cuál es el tamaño (forma) de nuestro DataFrame? ¿Podríamos considerarlo un dataset grande o pequeño?

R: (3033, 12), eso es grande dependiendo de que hagamos (si usamos una SVM digamos, si es grande), pero si, es grande.

3. ¿Crees que los nombres de las columnas son suficientemente descriptivos? ¿Podrían ser más claros o limpios?

R: Si, de hecho hasta tiene el tipo de dato.

4. ¿Qué tipos de datos tenemos?

R: Objeto (string), int, float (double)

5. ¿Qué significa el formato en el que tenemos las fechas?

R: The Unix Epoch, es una convención de usar el tiempo desde esa fecha en milisegundos según.

6. ¿Qué tipo de preguntas podríamos responder usando los datos numéricos de este dataset?

R: Que semanas hubo más bestsellers, y el precio de estos como varia.

7. ¿Qué tipo de preguntas podríamos responder usando los datos no-numéricos?

R: El autor, editorial que aparece más en la lista y así.